Sources/GRPCCodeGen/Internal/Renderer/TextBasedRenderer.swift

```diff
@@ -166,6 +166,7 @@ struct TextBasedRenderer: RendererProtocol {
       commentString = string
     }
     if prefix.isEmpty {
+      while commentString.hasSuffix("\n") { commentString.removeLast() }
```

Comment:
Where are the problematic extra new lines coming from with the preformatted comments? Modifying it here doesn't seem like the right thing to do, they are after all preformatted and we're re-formatting them...

Reply:
So if we have a proto that's like:

```proto
// Some comments about Foo
service Foo {
}
```

and get the comments using `protoSourceComments()`, do we get `"// Some comments about Foo"`, `"// Some comments about Foo\n"` or `"// Some comments about Foo\n\n`" or something else?

It's not clear to me where the extra newline is coming from but I don't think we should reformat pre-formatted comments.

Reply:
yes, we get "// Some comments about Foo\n"

Reply:
Okay, so how come when we render it we get the extra new line? Does the writer add a newline?

Reply:
The `writeLine()` function of the renderer that is used sets a property "nextWriteAppendsToLastLine" to false at the end, this way a new line is added after every line that is rendered.

Reply:
I'm not sure that's the problem, I think the issue is that `writeLine` just appends a single `String` to an array of `lines` in the writer which are eventually joined by newlines. In other words the writer adds in new lines between each "line" added by `writeLine`. 

However when we render the preformatted comments we append the whole block as a single line. The code below - for rendering the other comments - is different: it splits the lines and writes each separately. This lets the _writer_ deal with adding new lines.

I think we should do the same _but_ when the last line is empty (i.e. the text ended in "\n" before it was split) we should discard that empty line.